R/methods-Megastudy.R

```diff
@@ -0,0 +1,246 @@
+  #since we validate theyre all the same, can just take the first
+  return(veupathUtils::getEntityId(object[[1]]@variableSpec))
+})
+
+#' @include methods-Statistic.R
+#' @export
+setMethod('as.data.table', signature('StudySpecificVocabulary'), function(x) {
+  .dt <- data.table::data.table('study'=x@study, 'variable'=x@vocabulary)
+  names(.dt) <- c(x@studyIdColumnName, getColName(x@variableSpec))
+
+  return(.dt)
+})
+
+#' @export 
+setMethod('as.data.table', signature('StudySpecificVocabulariesByVariable'), function(x) {
+  return(purrr::reduce(lapply(as.list(x), veupathUtils::as.data.table), rbind))
+})
+
+#should this be an s4 method?
+findEntityIdColumnNameForVariableSpec <- function(varSpec, entityIdColumns) {
+  if (!inherits(varSpec, 'VariableSpec')) stop("The first argument must be of the S4 class `VariableSpec`.")
+
+  return(entityIdColumns[grepl(varSpec@entityId, entityIdColumns)])
+}
+
+findStudyVocabularyByVariableSpec <- function(vocabs, variableSpec) {
+  if (!inherits(vocabs, 'StudySpecificVocabulariesByVariableList')) stop("The first argument must be of the S4 class `StudySpecificVocabulariesByVariableList`.")
+  if (!inherits(variableSpec, 'VariableSpec')) stop("The second argument must be of the S4 class `VariableSpec`.")
+
+  vocabVariableSpecs <- lapply(as.list(vocabs), veupathUtils::getVariableSpec)
+  index <- which(purrr::map(vocabVariableSpecs, function(x) {veupathUtils::getColName(x)}) == veupathUtils::getColName(variableSpec))
+
+  return(vocabs[[index]])  
+}
+
+#' Impute Zeroes (on tall data)
+#' 
+#' This function returns a data.table which has explicit zero values
+#' for all expected categories of some variable of interest in a megastudy.
+#' 
+#' @param object veupathUtils::Megastudy
+#' @return data.table
+#' @include class-VariableMetadata.R
+#' @export
+setGeneric("getDTWithImputedZeroes", 
+  function(object, variables) standardGeneric("getDTWithImputedZeroes"),
+  signature = c("object", "variables")
+)
+
+#' @importFrom digest digest
+#' @export
+setMethod('getDTWithImputedZeroes', signature = c('Megastudy', 'VariableMetadataList'), function (object, variables) {
+ 
+  weightingVariablesMetadata <- findWeightingVariablesMetadata(variables)
+  if (is.null(weightingVariablesMetadata)) return(object@data)
+
+  .dt <- object@data
+  # TODO feel like im doing this operation a lot.. maybe another method/ helper?
+  # also, try to figure a way we dont have to do this.. i dont remember why i did this and its inconsistent behavior
+  variableColumnNames <- unlist(lapply(lapply(as.list(variables), veupathUtils::getVariableSpec), veupathUtils::getColName))
+  allEntityIdColumns <- object@ancestorIdColumns
+  # drop things that arent in the plot, except ids
+  .dt <- .dt[, c(variableColumnNames, allEntityIdColumns), with=FALSE]
+  vocabs <- object@studySpecificVocabularies
+
+  # it seems a lot of this validation could belong to some custom obj w both a megastudy and vm slot.. but what is that? a MegastudyPlot?
```

Comment:
we could. i thought about it for a second as well.. i think part of the reason i hesitated is bc im not sure where this will go once we add other plan d features. but i suppose a validation fxn isnt too much work and equally easy to refactor later.